### PR TITLE
Add drone pipeline for building Connect with signed tsh.app

### LINKED ISSUE #15681: Add drone pipeline for building Connect with signed tsh.app
To build a release version of Connect, we need:

* a signed version of tsh.app
* Node.js
* the developer ID cert from Keychain

darwinTagPipeline has Node.js but at that point tsh.app isn’t signed yet. darwinPkgPipeline signs tsh.app but doesn’t have Node.js.

## Where we are now

**pipeline**|**artifacts**|**toolchains**|**notarization & home dirs**
-----|-----|-----|-----
darwinTagPipeline|publishes Teleport binaries & Connect|needs Node, Rust, and Go toolchains|notarizes Connect, needs HOME to be in /Users
darwinPkgPipeline|downloads artifacts made by the tag pipeline, publishes .pkg & tsh-only .pkg|doesn’t need any toolchains|notarizes tsh.app, needs HOME to be in /Users
darwinPushPipeline|builds Teleport binaries & Connect without publishing|needs Node, Rust, and Go toolchains|doesn’t notarize and doesn’t need HOME to be changed; still, it installs toolchains in the same way as darwinTagPipeline

## Where we want to be

**pipeline**|**artifacts**|**toolchains**|**notarization & home dirs**
-----|-----|-----|-----
darwinTagPipeline|🛑 publishes Teleport binaries|🛑 needs Rust and Go toolchains|🛑 doesn’t notarize, doesn’t need special accommodations for HOME
darwinPkgPipeline|downloads artifacts made by the tag pipeline, publishes .pkg & tsh-only .pkg|doesn’t need any toolchains|notarizes tsh.app, needs HOME to be in /Users
darwinPushPipeline|builds Teleport binaries & Connect without publishing|needs Node, Rust, and Go toolchains|doesn’t notarize and doesn’t need HOME to be changed
darwinConnectDmgPipeline|downloads tsh-only .pkg, publishes Connect .dmg|needs Node toolchain|notarizes Connect, needs HOME to be in /Users

What follows from those tables is that only darwinTagPipeline would require some changes. We’d mostly remove responsibilities from it and move them to darwinConnectDmgPipeline.

At the moment though, both darwinTagPipeline and darwinPushPipeline have special HOME handling to support notarizing & toolchains in temporary workspaces, as described in 86f3a3d618. However, after the changes described here none of those two pipelines would notarize anything, so technically they could go back to the original HOME handling. But, darwinPushPipeline and darwinConnectDmgPipeline will need to share the steps needed to build Connect.

darwinConnectDmgPipeline can acquire tsh.app by downloading the tsh-only .pkg and calling `pkgutil --expand-full`.

### PR DESCRIPTION
**Pull request by ravicious**
I'd like to include this change in 10.2 so I'd appreciate if you could take a look at this PR sooner later than later, if possible.

Normally I'd first verify if the changes work but [drone is down at the moment](https://gravitational.slack.com/archives/C0DF0TPMY/p1661244379021479?thread_ts=1661243199.148379&cid=C0DF0TPMY) and doesn't make new builds. Thus, in lieu of a speedy review, I'm submitting this PR already. I'll verify that it actually works as soon as possible. FWIW, the first commit from this branch [built the app without problems](https://drone.platform.teleport.sh/gravitational/teleport/14880/4/6) but crashed on copying artifacts, I have already fixed this issue since.

Closes #15681, best reviewed commit-by-commit.

# Why

A more detailed plan is described in #15681 (with a nice table and all). But the gist of it is that the tag pipeline right now builds a Connect .app bundle that includes a standalone tsh binary.

However, in order to support Touch ID and share Secure Enclave keys between tsh.app and Connect (gravitational/webapps#1116) we need to bundle not a standalone binary but the binary wrapped in an .app bundle that has been signed and has an appropriate provisioning profile (this is an Apple term).

# How

In our case, this is as simple as copying signed tsh.app within the Connect bundle rather than copying a standalone binary.

Since tsh.app is built and signed only within the `build-darwin-amd64-pkg-tsh` pipeline, I added a new pipeline which waits for `build-darwin-amd64-pkg-tsh` to finish and upload its artifacts to S3. After it is done, my new pipeline downloads tsh.pkg, unpacks it, grabs tsh.app and includes it in Connect's app bundle.

This is similar to how `build-darwin-amd64-pkg-tsh` already works by waiting for `build-darwin-amd64` to upload binaries to S3 which are then packaged into .pkg files.

In the end, we end up with three pipelines, created by calling `darwinTagPipeline`, `darwinPushPipeline`, `darwinConnectDmgPipeline, which share code by calling the same functions but with different options since all three pipelines have slightly different needs (see the second table in #15681).

# How but with more details

I also refactored a bunch of code since `build-darwin-amd64` no longer needed to build Connect nor sign & notarize anything. So in first commit I push a proof of concept that includes a lot of copy-pasted code which then gets cleaned up in subsequent commits.

For more details on how the darwin pipelines work and what changes were needed to accommodate for this new process, see the tables in #15681.

### CODE REVIEW

File: dronegen/mac.go [hunk 1]
```diff
@@ -365,19 +465,59 @@ func darwinTagBuildCommands(b buildType, opts darwinBuildOptions) []string {
...
 
-	if b.hasTeleportConnect() {
+	// TODO: Remove building Connect and Node stuff from tag pipeline. However, these are currently
+	// shared between push and tag. Push will have to continue building Connect without signing.
+	if opts.hasTeleportConnect {
 		commands = append(commands,
 			`cd $WORKSPACE_DIR/go/src/github.com/gravitational/webapps`,
 			// c.extraMetadata.version overwrites the version property from package.json to $VERSION
 			// https://www.electron.build/configuration/configuration.html#Configuration-extraMetadata
-			`yarn install --frozen-lockfile && yarn build-term && yarn package-term -c.extraMetadata.version=$VERSION`,
```
Comment (ravicious) [inline dronegen/mac.go:373]: Removed because it has no effect due to how the webapps repo is set up, see gravitational/webapps#1047.

File: dronegen/mac.go [hunk 2]
```diff
@@ -517,34 +517,28 @@ func darwinConnectBuildCommands() []string {
 	return commands
 }
 
-func darwinTagCopyPackageArtifactCommands(b buildType, hasTeleportConnect bool) []string {
+func darwinTagCopyPackageArtifactCommands() []string {
```
Comment (ravicious) [inline dronegen/mac.go:599]: `darwinTagPipeline` no longer builds the Connect artifact so this function no longer needs to care about that artifact.

File: dronegen/mac.go [hunk 3]
```diff
@@ -43,8 +43,81 @@ func newDarwinPipeline(name string) pipeline {
...
+				"AWS_ACCESS_KEY_ID":     {fromSecret: "AWS_ACCESS_KEY_ID"},
+				"AWS_SECRET_ACCESS_KEY": {fromSecret: "AWS_SECRET_ACCESS_KEY"},
+				"AWS_REGION":            {raw: "us-west-2"},
+				"WORKSPACE_DIR":         {raw: p.Workspace.Path},
+			},
+			Commands: darwinUploadToS3Commands(),
+		},
+		{
+			Name:     "Register artifact",
+			Commands: tagCreateReleaseAssetCommands(b, "", nil),
```
Comment (codingllama) [inline dronegen/mac.go:101]: nit: What do `""` and `nil` mean? Please add an inline comment.

**Review by codingllama — APPROVED** [review]
Seems reasonable to me, although Drone tends to be the decider on whether the pipelines actually work.  ;)

My preference would be for less interdependent stages, so instead of adding a new pipeline I'd rather just do everything needed for Connect in the same stage where we get the required tsh.app. I find the pipelines we have hard to follow/reason about, I'd much rather just have a shell script that does it all in one go. (Not a blocker, obviously.)

File: dronegen/mac.go [hunk 4]
```diff
@@ -277,28 +349,78 @@ func installNodeToolchainStep(workspacePath string) step {
 	}
 }
 
-func cleanUpToolchainsStep(path string) step {
-	return step{
+func configureToolchainsCommands(config toolchainConfig) []string {
+	commands := []string{
+		// HOME needs to be set to the actual user rather than what Drone sets it to by default. This
+		// way we're able to unlock Keychain which is needed for Connect signing.
+		`echo HOME=$${HOME}`,
```
Comment (jakule) [inline dronegen/mac.go:356]: Is this `echo` needed for some reason, or it's here just for debug?

**Review by jakule — APPROVED** [review]

File: dronegen/mac.go (hunk 4 above)
Comment (ravicious) [reply dronegen/mac.go:356]: I think it might be a leftover from #12257 when Trent was trying to make signing work (688a3712b). It doesn't seem to serve any purpose anymore as it just shows what Drone has set HOME to and the var gets immediately overwritten by the next command. I'll remove it.

**Comment by ravicious** [comment]
@codingllama 

> My preference would be for less interdependent stages, so instead of adding a new pipeline I'd rather just do everything needed for Connect in the same stage where we get the required tsh.app. I find the pipelines we have hard to follow/reason about, I'd much rather just have a shell script that does it all in one go. (Not a blocker, obviously.)

I looked at previous PRs which were adding Connect-related stuff to dronegen and it looks like Walt has an exactly opposite opinion, see https://github.com/gravitational/teleport/pull/12257#pullrequestreview-954183021 😅

I thought of building Connect in the same pipeline that we make tsh.app. At the time the whole toolchain setup looked quite complex to me so I didn't want to pollute `build-darwin-amd64-pkg-tsh` with toolchain steps.

I think the nice thing about this solution is that even if the Connect build fails, all the previous artifacts get build which is what 90% of devs care about. And I consider the build of Connect to be the most volatile due to it being steeped in the Node.js ecosystem.

**Comment by ravicious** [comment]
Got [a passing build](https://drone.platform.teleport.sh/gravitational/teleport/14938).

**Comment by codingllama** [comment]
> I looked at previous PRs which were adding Connect-related stuff to dronegen and it looks like Walt has an exactly opposite opinion, see [#12257 (review)](https://github.com/gravitational/teleport/pull/12257#pullrequestreview-954183021) 😅

The whole pushing to AWS to then immediately download from AWS, take the packages apart and reassemble seems overly complicated to me, but no worries. Let's move on.

> I think the nice thing about this solution is that even if the Connect build fails, all the previous artifacts get build which is what 90% of devs care about. And I consider the build of Connect to be the most volatile due to it being steeped in the Node.js ecosystem.

A broken pipeline is a blocked release, no matter how much of it worked, so I'm not sure that matters much. (The Node.js ecosystem _is_ concerning, I grant you that.)
